File: sdks/javascript-admin/src/cluster.ts
/**
 * Cluster service for managing nodes, storage, settings, and monitoring.
 */

import type { RequestFn } from './index';

// ============================================================================
// Types
// ============================================================================

export interface ClusterInfo {
  cluster_name: string;
  node_count: number;
  leader_id?: string;
  status: string;
  nodes: ClusterNode[];
}

export interface ClusterNode {
  id: string;
  name: string;
  address: string;
  port: number;
  role: string;
  status: string;
  last_heartbeat?: string;
  uptime_seconds?: number;
}

export interface NodeLogs {
  node_id: string;
  logs: LogEntry[];
}

export interface LogEntry {
  timestamp: string;
  level: string;
  message: string;
  target?: string;
}

export interface StorageInfo {
  total_bytes: number;
  used_bytes: number;
  available_bytes: number;
  tables: StorageTableInfo[];
}

export interface StorageTableInfo {
  name: string;
  size_bytes: number;
  row_count: number;
}

export interface DatabaseStats {
  total_queries: number;
  active_connections: number;
  uptime_seconds: number;
  tables_count: number;
  total_rows: number;
  cache_hit_ratio?: number;
  avg_query_time_ms?: number;
}

export interface ServerStats {
  cpu_usage?: number;
  memory_usage_bytes?: number;
  memory_total_bytes?: number;
  disk_usage_bytes?: number;
  disk_total_bytes?: number;
  open_connections: number;
  total_requests: number;
  requests_per_second?: number;
}

export interface Alert {
  id: string;
  severity: string;
  message: string;
  source: string;
  timestamp: string;
  acknowledged: boolean;
}

export interface Activity {
  id: string;
  action: string;
  user?: string;
  resource?: string;
  timestamp: string;
  details?: Record<string, unknown>;
}

export interface Settings {
  [key: string]: unknown;
}

// ============================================================================
// Service
// ============================================================================

export class ClusterService {
  constructor(private request: RequestFn) {}

  /** Get cluster overview. */
  async getCluster(): Promise<ClusterInfo> {
    return this.request<ClusterInfo>('GET', '/api/v1/admin/cluster');
  }

  /** List all nodes. */
  async listNodes(): Promise<ClusterNode[]> {
    return this.request<ClusterNode[]>('GET', '/api/v1/admin/nodes');
  }

  /** Restart a node. */
  async restartNode(nodeId: string): Promise<{ message: string }> {
    return this.request<{ message: string }>('POST', `/api/v1/admin/nodes/${encodeURIComponent(nodeId)}/restart`);
  }

  /** Drain a node (stop accepting new connections and migrate work). */
  async drainNode(nodeId: string): Promise<{ message: string }> {
    return this.request<{ message: string }>('POST', `/api/v1/admin/nodes/${encodeURIComponent(nodeId)}/drain`);
  }

  /** Get logs for a specific node. */
  async getNodeLogs(nodeId: string): Promise<NodeLogs> {
    return this.request<NodeLogs>('GET', `/api/v1/admin/nodes/${encodeURIComponent(nodeId)}/logs`);
  }

  /** Remove a node from the cluster. */
  async removeNode(nodeId: string): Promise<void> {
    await this.request<void>('DELETE', `/api/v1/admin/nodes/${encodeURIComponent(nodeId)}`);
  }

  /** Get storage information. */
  async getStorage(): Promise<StorageInfo> {
    return this.request<StorageInfo>('GET', '/api/v1/admin/storage');
  }

  /** Get server stats (CPU, memory, connections). */
  async getStats(): Promise<ServerStats> {
    return this.request<ServerStats>('GET', '/api/v1/admin/stats');
  }

  /** Get database stats (queries, tables, rows). */
  async getDatabase(): Promise<DatabaseStats> {
    return this.request<DatabaseStats>('GET', '/api/v1/admin/database');
  }

  /** Get active alerts. */
  async getAlerts(): Promise<Alert[]> {
    return this.request<Alert[]>('GET', '/api/v1/admin/alerts');
  }

  /** Get recent activity log. */
  async getActivities(): Promise<Activity[]> {
    return this.request<Activity[]>('GET', '/api/v1/admin/activities');
  }

  /** Get server settings. */
  async getSettings(): Promise<Settings> {
    return this.request<Settings>('GET', '/api/v1/admin/settings');
  }

  /** Update server settings. */
  async updateSettings(settings: Settings): Promise<Settings> {
    return this.request<Settings>('PUT', '/api/v1/admin/settings', settings);
  }
}

File: sdks/python-admin/aegis_db_admin/cluster.py
"""
Aegis-DB Admin SDK - Cluster Service

Cluster, node, storage, stats, alerts, activities, and settings management.

@version 1.0.0
@author AutomataNexus Development Team
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, TYPE_CHECKING

from .types import (
    ClusterInfo,
    NodeInfo,
    StorageInfo,
    QueryStats,
    DatabaseStats,
    Alert,
    Activity,
)

if TYPE_CHECKING:
    from .client import AegisAdmin


class ClusterService:
    """Manage cluster, nodes, settings, and observability."""

    def __init__(self, admin: AegisAdmin) -> None:
        self._admin = admin

    # =========================================================================
    # Cluster
    # =========================================================================

    async def get_cluster_info(self) -> ClusterInfo:
        """Get cluster information.

        Returns:
            ClusterInfo object.
        """
        data = await self._admin._request("GET", "/api/v1/admin/cluster")
        return ClusterInfo(
            cluster_name=data.get("cluster_name"),
            node_count=data.get("node_count", 0),
            leader=data.get("leader"),
            status=data.get("status"),
            version=data.get("version"),
            _raw=data,
        )

    # =========================================================================
    # Nodes
    # =========================================================================

    async def list_nodes(self) -> List[NodeInfo]:
        """List all cluster nodes.

        Returns:
            List of NodeInfo objects.
        """
        data = await self._admin._request("GET", "/api/v1/admin/nodes")
        nodes_list = data if isinstance(data, list) else data.get("nodes", [])
        return [_parse_node(n) for n in nodes_list]

    async def restart_node(self, node_id: str) -> Dict[str, Any]:
        """Restart a cluster node.

        Args:
            node_id: The node identifier.

        Returns:
            Server response dict.
        """
        return await self._admin._request(
            "POST", f"/api/v1/admin/nodes/{node_id}/restart"
        )

    async def drain_node(self, node_id: str) -> Dict[str, Any]:
        """Drain a cluster node (stop accepting new work).

        Args:
            node_id: The node identifier.

        Returns:
            Server response dict.
        """
        return await self._admin._request(
            "POST", f"/api/v1/admin/nodes/{node_id}/drain"
        )

    async def get_node_logs(
        self,
        node_id: str,
        *,
        lines: Optional[int] = None,
        level: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get logs from a node.

        Args:
            node_id: The node identifier.
            lines: Number of log lines to retrieve.
            level: Log level filter (e.g. 'error', 'warn', 'info').

        Returns:
            Dict containing log entries.
        """
        params: List[str] = []
        if lines is not None:
            params.append(f"lines={lines}")
        if level is not None:
            params.append(f"level={level}")
        qs = f"?{'&'.join(params)}" if params else ""
        return await self._admin._request(
            "GET", f"/api/v1/admin/nodes/{node_id}/logs{qs}"
        )

    async def remove_node(self, node_id: str) -> Dict[str, Any]:
        """Remove a node from the cluster.

        Args:
            node_id: The node identifier.

        Returns:
            Server response dict.
        """
        return await self._admin._request(
            "DELETE", f"/api/v1/admin/nodes/{node_id}"
        )

    # =========================================================================
    # Storage / Stats / Database
    # =========================================================================

    async def get_storage_info(self) -> StorageInfo:
        """Get storage information.

        Returns:
            StorageInfo object.
        """
        data = await self._admin._request("GET", "/api/v1/admin/storage")
        return StorageInfo(
            total_bytes=data.get("total_bytes", 0),
            used_bytes=data.get("used_bytes", 0),
            free_bytes=data.get("free_bytes", 0),
            backend=data.get("backend"),
            _raw=data,
        )

    async def get_query_stats(self) -> QueryStats:
        """Get query statistics.

        Returns:
            QueryStats object.
        """
        data = await self._admin._request("GET", "/api/v1/admin/stats")
        return QueryStats(
            total_queries=data.get("total_queries", 0),
            queries_per_second=data.get("queries_per_second", 0.0),
            avg_latency_ms=data.get("avg_latency_ms", 0.0),
            _raw=data,
        )

    async def get_database_stats(self) -> DatabaseStats:
        """Get database statistics.

        Returns:
            DatabaseStats object.
        """
        data = await self._admin._request("GET", "/api/v1/admin/database")
        return DatabaseStats(
            table_count=data.get("table_count", 0),
            total_rows=data.get("total_rows", 0),
            total_size_bytes=data.get("total_size_bytes", 0),
            _raw=data,
        )

    # =========================================================================
    # Alerts / Activities
    # =========================================================================

    async def get_alerts(self) -> List[Alert]:
        """Get active alerts.

        Returns:
            List of Alert objects.
        """
        data = await self._admin._request("GET", "/api/v1/admin/alerts")
        alerts_list = data if isinstance(data, list) else data.get("alerts", [])
        return [
            Alert(
                id=a.get("id"),
                severity=a.get("severity"),
                message=a.get("message"),
                timestamp=a.get("timestamp"),
                acknowledged=a.get("acknowledged", False),
                _raw=a,
            )
            for a in alerts_list
        ]

    async def get_activities(self) -> List[Activity]:
        """Get activity log.

        Returns:
            List of Activity objects.
        """
        data = await self._admin._request("GET", "/api/v1/admin/activities")
        acts_list = data if isinstance(data, list) else data.get("activities", [])
        return [
            Activity(
                id=a.get("id"),
                action=a.get("action"),
                user=a.get("user"),
                timestamp=a.get("timestamp"),
                details=a.get("details"),
                _raw=a,
            )
            for a in acts_list
        ]

    # =========================================================================
    # Settings
    # =========================================================================

    async def get_settings(self) -> Dict[str, Any]:
        """Get current server settings.

        Returns:
            Settings dict.
        """
        return await self._admin._request("GET", "/api/v1/admin/settings")

    async def update_settings(self, settings: Dict[str, Any]) -> Dict[str, Any]:
        """Update server settings.

        Args:
            settings: Dict of settings to update.

        Returns:
            Updated settings dict.
        """
        return await self._admin._request("PUT", "/api/v1/admin/settings", settings)


# =============================================================================
# Helpers
# =============================================================================

def _parse_node(data: Dict[str, Any]) -> NodeInfo:
    return NodeInfo(
        id=data.get("id"),
        name=data.get("name"),
        address=data.get("address"),
        role=data.get("role"),
        status=data.get("status"),
        uptime=data.get("uptime"),
        version=data.get("version"),
        _raw=data,
    )
